### PR TITLE
Support for extension attributes

### LINKED ISSUE #112: Feature Request: Make the extension aware of Magento's extension attributes
Come up with a plan to make the PHPStan extension aware of defined [extension attributes](https://devdocs.magento.com/guides/v2.4/extension-dev-guide/attributes.html#extension).

### PR DESCRIPTION
**Pull request by mattwellss**
Resolves #112

### CODE REVIEW

**Comment by shochdoerfer** [comment]
Thanks a lot for your hard work on this. Very much appreciated! It'll take some time until I can check your PR, I am quite busy these days.

One thing we should do is to add `laminas/laminas-code` as a composer dependency if possible. Technically, the PHPStan extension is supporting anything `magento/framework` >101.0.0 (Magento 2.2.x IIRC). Back then dependency was called `zendframework/zend-code` which would break your implementation. We would need to check with which version constraints `laminas/laminas-code` is included in the different Magento versions and then come up with needed Composer constraints.

Is this PR blocked by #161 or could we treat that independently?

**Comment by shochdoerfer** [comment]
@sprankhub can you help out testing the changes, please? Since you seem to have some projects that make use of extension attributes, it would be lovely to get some real world feedback. I'd need to build a quick module myself to test things.

**Comment by mattwellss** [comment]
> One thing we should do is to add laminas/laminas-code as a composer dependency if possible. Technically, the PHPStan extension is supporting anything magento/framework >101.0.0 (Magento 2.2.x IIRC). Back then dependency was called zendframework/zend-code which would break your implementation.

Good point. I'll check it out.

> Is this PR blocked by #161 or could we treat that independently?

Not blocked, I only noticed it because in my own Magento environment I've configured PHPStan's cache to be project-local.

**Comment by sprankhub** [comment]
Well, actually it was not me, but my colleague :D I'll see what we can do regarding testing :)

**Comment by shochdoerfer** [comment]
@mattwellss will take a deeper dive in this next week. Need to finish a few projects at work before my holiday next week. Should have a bit more time then. I try to get the next release including this PR done before Christmas in the best case.

**Comment by mattwellss** [comment]
Great question, @shochdoerfer. I'll take a look tomorrow... I can see better now your concerns about the extension being able to find extension attributes for "non-application" usage.

File: tests/bitExpert/PHPStan/Magento/Autoload/ExtensionInterfaceAutoloaderUnitTest.php [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        $autoloader = new ExtensionInterfaceAutoloader(
+            new Cache(new FileCacheStorage($this->root->url() . '/tmp/cache/PHPStan')),
+            $this->root->url()
+        );
+
+        require $this->root->url() . '/interface.php';
+
+        $autoloader->autoload($interfaceName);
+        static::assertTrue(interface_exists($interfaceName));
+        $interfaceReflection = new \ReflectionClass($interfaceName);
```
Comment (mattwellss) [inline tests/bitExpert/PHPStan/Magento/Autoload/ExtensionInterfaceAutoloaderUnitTest.php:114]: I'm unsure how to resolve this error (outside of suppressing it). PHPStan definitely cannot find this class using static analysis!

Comment (shochdoerfer) [reply tests/bitExpert/PHPStan/Magento/Autoload/ExtensionInterfaceAutoloaderUnitTest.php:114]: I am fine if we ignore the error. I mean me have to make some magic here, that's fine then :)

**Comment by shochdoerfer** [comment]
@mattwellss did a quick check and things look good. Works even for stand-alone modules, it seems. Tested it against `magento/module-inventory-api`. With the `ExtensionInterfaceAutoloader` not being active, I get quite a few extension interface errors. After enabling the autoloader the errors are gone.

The only thing that we need to change is the `laminas-code` version as older Magento versions require version 3.3 (Magento 2.3.6) or 3.4.1 (Magento 2.3.7). I'll check how "compatible" the different versions are and fix them before merging your changes.

**Comment by mattwellss** [comment]
> did a quick check and things look good. Works even for stand-alone modules, it seems.

That's great! I've been a bit inundated with time-sensitive tasks, so I haven't been able to complete a test.

I was planning to test the efficacy of both core extension attributes and custom ones, which I imagine might fail. If you were able to verify that both are fine then I can skip it.

I'll correct the version requirement for laminas code this week.

**Comment by shochdoerfer** [comment]
> I was planning to test the efficacy of both core extension attributes and custom ones, which I imagine might fail. If you were able to verify that both are fine then I can skip it.

Would be cool if you could check that. I haven't done much so far with extension attributes, I don't want to miss something important :)

> I'll correct the version requirement for laminas code this week.

Not needed, I am currently running some tests and have fixed it locally. Will commit it once I manually merge the PR.

**Comment by shochdoerfer** [comment]
I was hoping that it would be possible to install `zendframework/zend-code` and `laminas/laminas-code` at the same time. Sadly, that is not possible. That means we need up raise the minimum requirement for `magento/framework` to 102.0.0 (Magento 2.3.0). Fine for me, not sure if anyone is still using this extension with a Magento 2.2 instance. And if so, they can still stick to an older version of the extension.

**Pull request merged**